### PR TITLE
adding merge functionality for eps to regular appointments

### PR DESCRIPTION
**Pull request by lee-delarm6**
## Summary

- This work is behind a param
- This change introduces EPS (Electronic Prescribing System) appointment integration into the VAOS (VA Online Scheduling) system.
- Implementing EPS appointment normalization and merging functionality
- Modifying appointment handling logic to incorporate EPS appointments
- Cleaning up some code formatting and rubocop issues
- The solution integrates EPS appointments with existing VAOS appointments by normalizing the EPS appointment data structure to match VAOS format and merging them while avoiding duplicates. This unified approach allows for consistent appointment handling across both systems.
- Based on the code patterns and feature toggles present, this appears to be a gradual integration approach allowing for controlled rollout and testing.

## Testing done

- [x] *New code is covered by unit tests*
- New behavior: Both VAOS and EPS appointments are merged and displayed when the EPS include flag is present
- Run this function with an without EPS appointments

## What areas of the site does it impact?
- VAOS (VA Online Scheduling) system, specifically:

Appointment retrieval and display
Configuration settings
Appointment service layer

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [x]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x]  Feature/bug has a monitor built into Datadog (if applicable)
- [x]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

### CODE REVIEW

File: modules/vaos/app/controllers/vaos/v2/appointments_controller.rb [hunk 1]
```diff
@@ -92,6 +99,11 @@ def appointments
           appointments_service.get_appointments(start_date, end_date, statuses, pagination_params, include_index_params)
       end
 
+      def eps_appointments
+        @eps_appointments ||=
+          eps_appointments_service.get_appointments(patient_id: Settings.vaos.eps.fake_patient_id)
```
Comment (randomsync) [inline modules/vaos/app/controllers/vaos/v2/appointments_controller.rb:104]: why `fake_patient_id`?

File: modules/vaos/app/controllers/vaos/v2/appointments_controller.rb [hunk 2]
```diff
@@ -21,14 +21,16 @@ class AppointmentsController < VAOS::BaseController
       COMMENT = 'comment'
 
       def index
-        appointments[:data].each do |appt|
+        merged_appointments = appointments_service.merge_appointments(eps_appointments, appointments)
```
Comment (randomsync) [inline modules/vaos/app/controllers/vaos/v2/appointments_controller.rb:24]: I'd suggest putting these changes behind a feature flag

File: modules/vaos/app/services/vaos/v2/appointments_service.rb [hunk 3]
```diff
@@ -204,6 +204,54 @@ def get_facility_timezone_memoized(facility_location_id)
...
+
+      # rubocop:enable Lint/DuplicateBranch
+
+      def calculate_end_time(start_time)
+        return nil unless start_time
+
+        Time.zone.parse(start_time) + 60.minutes
+      end
+
+      def merge_appointments(new_appointments, basic_appointments)
```
Comment (randomsync) [inline modules/vaos/app/services/vaos/v2/appointments_service.rb:245]: I feel that 'new_appointments' and 'basic_appointments' might lead to confusion as to what makes an appt a new vs a basic. Would it make sense to call them what they are, which is eps_appointments and vaos_appointments?

File: modules/vaos/app/services/vaos/v2/appointments_service.rb [hunk 4]
```diff
@@ -204,6 +204,54 @@ def get_facility_timezone_memoized(facility_location_id)
...
+        normalized_new = new_appointments[:appointments].map { |appt| normalize_eps_appointment(appt) }
+        basic_data = basic_appointments[:data].is_a?(Array) ? basic_appointments[:data] : [basic_appointments[:data]]
+        existing_ids = basic_data.to_set { |a| a[:referralId] }
+        merged_data = basic_data + normalized_new.reject { |a| existing_ids.include?(a[:referralId]) }
```
Comment (randomsync) [inline modules/vaos/app/services/vaos/v2/appointments_service.rb:249]: are we deduping by referralId?

File: modules/vaos/app/controllers/vaos/v2/appointments_controller.rb [hunk 5]
```diff
@@ -21,14 +21,16 @@ class AppointmentsController < VAOS::BaseController
...
+
+        merged_appointments.each do |appt|
```
Comment (randomsync) [inline modules/vaos/app/controllers/vaos/v2/appointments_controller.rb:26]: `appointments_service.merge_appointments` is returning a hash of the form ` { data: [... appts array] }`. Is this supposed to iterate over the appts array?

File: modules/vaos/app/controllers/vaos/v2/appointments_controller.rb [hunk 6]
```diff
@@ -21,14 +21,16 @@ class AppointmentsController < VAOS::BaseController
...
           set_facility_error_msg(appt) if include_index_params[:facilities]
           scrape_appt_comments_and_log_details(appt, index_method_logging_name, PAP_COMPLIANCE_TELE)
           log_appt_creation_time(appt)
         end
 
         serializer = VAOS::V2::VAOSSerializer.new
-        serialized = serializer.serialize(appointments[:data], 'appointments')
+        serialized = serializer.serialize(merged_appointments, 'appointments')
 
         if !appointments[:meta][:failures]&.empty?
```
Comment (randomsync) [inline modules/vaos/app/controllers/vaos/v2/appointments_controller.rb:35]: You'd need to update other code within this method that refers to original appointments array as well

**Review by randomsync — CHANGES_REQUESTED** [review]
I understand that this is a draft so just providing general feedback. There are other changes that would be needed to make sure the merged appointments are used in the rest of the controller method. Please add tests as well so it's easier to understand the changes and the desired effect.

File: modules/vaos/app/controllers/vaos/v2/appointments_controller.rb (hunk 6 above)
Comment (lee-delarm6) [reply modules/vaos/app/controllers/vaos/v2/appointments_controller.rb:35]: Note to self, appointments isn't called at all here

File: modules/vaos/app/services/vaos/v2/appointments_service.rb (hunk 4 above)
Comment (lee-delarm6) [reply modules/vaos/app/services/vaos/v2/appointments_service.rb:249]: Yes, just marking this as a confirmation

File: modules/vaos/app/services/vaos/v2/appointments_service.rb (hunk 3 above)
Comment (lee-delarm6) [reply modules/vaos/app/services/vaos/v2/appointments_service.rb:245]: To keep renaming to a minimum, I'm going to change it to `eps_appointments` and change `basic_appointments` back to appointments

File: modules/vaos/app/controllers/vaos/v2/appointments_controller.rb (hunk 1 above)
Comment (lee-delarm6) [reply modules/vaos/app/controllers/vaos/v2/appointments_controller.rb:104]: To test out with local data, I'll change to patient_id

File: modules/vaos/app/services/vaos/v2/appointments_service.rb [hunk 7]
```diff
@@ -204,6 +204,54 @@ def get_facility_timezone_memoized(facility_location_id)
...
+          'proposed'
+        end
+      end
+
+      # rubocop:enable Lint/DuplicateBranch
+
+      def calculate_end_time(start_time)
+        return nil unless start_time
+
+        Time.zone.parse(start_time) + 60.minutes
```
[the remark this answers is not in the document]
Comment (lee-delarm6) [reply modules/vaos/app/services/vaos/v2/appointments_service.rb:233]: I'm using the start and end from the same data point though, shouldn't this match up? This is from where the method is entered:

```
start: appt.dig(:appointmentDetails, :start),
end: calculate_end_time(appt.dig(:appointmentDetails, :start))
```

**Comment by randomsync** [comment]
@lee-delarm6 is this ready for review? there are no specs, and some of the other feedback is also un-addressed

File: modules/vaos/app/services/vaos/v2/appointments_service.rb [hunk 8]
```diff
@@ -827,6 +867,16 @@ def validate_response_schema(response, contract_name)
...
+      def eps_appointments_service
+        @eps_appointments_service ||=
+          Eps::AppointmentService.new(current_user)
+      end
+
+      def eps_appointments
+        @eps_appointments ||=
+          eps_appointments_service.get_appointments
+      end
     end
```
Comment (kanchanasuriya) [inline modules/vaos/app/services/vaos/v2/appointments_service.rb:880]: Isn't this supposed to be in the controller for fetching EPS appointments?  Did it get included in the PR accidentally?

Comment (lee-delarm6) [reply modules/vaos/app/services/vaos/v2/appointments_service.rb:880]: We moved all the logic to the service level, and we're passing in "eps" in the includes object instead of relying on the controller level now

File: modules/vaos/app/controllers/vaos/v2/appointments_controller.rb (hunk 5 above)
Comment (lee-delarm6) [reply modules/vaos/app/controllers/vaos/v2/appointments_controller.rb]: No longer does this (all controller logic removed)

File: modules/vaos/app/services/vaos/v2/appointments_service.rb [hunk 9]
```diff
@@ -204,6 +209,40 @@ def get_facility_timezone_memoized(facility_location_id)
...
+          locationId: appt[:locationId],
+          clinic: appt[:clinic],
+          contact: appt[:contact]
+        }.compact
+      end
+
+      def calculate_end_time(start_time)
+        return nil unless start_time
+
+        Time.zone.parse(start_time) + 60.minutes
```
[the remark this answers is not in the document]
Comment (lee-delarm6) [reply modules/vaos/app/services/vaos/v2/appointments_service.rb:233]: There is no end time, so we are making this assumption for prototype. End time MIGHT come later according to Stephen

File: modules/vaos/app/services/vaos/v2/appointments_service.rb [hunk 10]
```diff
@@ -204,6 +209,40 @@ def get_facility_timezone_memoized(facility_location_id)
...
+      end
+
+      def merge_appointments(new_appointments, basic_appointments)
+        normalized_new = new_appointments[:appointments].map { |appt| normalize_eps_appointment(appt) }
+        basic_data = basic_appointments[:data].is_a?(Array) ? basic_appointments[:data] : [basic_appointments[:data]]
```
[the remark this answers is not in the document]
Comment (lee-delarm6) [reply modules/vaos/app/services/vaos/v2/appointments_service.rb]: I verified with Simi that this will not break any other logic (and nothing in the controller) when pushing back a single item in an array (not for frontend either)

File: modules/vaos/app/services/vaos/v2/appointments_service.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (lee-delarm6) [reply modules/vaos/app/services/vaos/v2/appointments_service.rb:233]: This timezone is in UTC we have found, so we will also inform the front end to change it to the user's time zone

File: modules/vaos/app/services/vaos/v2/appointments_service.rb (hunk 8 above)
Comment (lee-delarm6) [reply modules/vaos/app/services/vaos/v2/appointments_service.rb:880]: Changed current_user to user

**Review by devin-mccurdy — DISMISSED** [review]
LGTM

File: modules/vaos/spec/services/v2/appointment_service_spec.rb [hunk 11]
```diff
@@ -1052,6 +1111,20 @@
     end
   end
 
+  describe '#get_appointments merge' do
+    context 'when include eps is true' do
+      it 'merges eps appointments with vaos appointments' do
+        VCR.use_cassette('vaos/eps/get_appointments_200_with_merge',
+                         match_requests_on: %i[method path query], allow_playback_repeats: true, tag: :force_utf8) do
+          allow_any_instance_of(Eps::AppointmentService).to receive(:get_appointments).and_return(eps_appointments)
```
Comment (randomsync) [inline modules/vaos/spec/services/v2/appointment_service_spec.rb:1119]: curious why we're mixing use of VCR as well as a mock

File: modules/vaos/app/serializers/vaos/v2/eps_appointment_serializer.rb [hunk 12]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+module VAOS
+  module V2
+    class EpsAppointmentSerializer
+      def serialize(appt)
+        {
+          id: appt[:id].to_s,
+          status: appt[:state] == 'booked',
```
Comment (randomsync) [inline modules/vaos/app/serializers/vaos/v2/eps_appointment_serializer.rb:9]: are we sure we want this? if appt[:state] == 'booked', then should the status be true? and vice-versa?

**Review by randomsync — CHANGES_REQUESTED** [review]
I think this PR still needs 2 things:
1. make sure your spec's mock response matches the response schema, otherwise after serialization most of the fields will be nil
2. currently, the spec doesn't validate the de-dupe functionality since the appointment response vcr cassette has no appointments with referral ids. we should add more specs to validate those cases (it might be easier to put all this in a new spec file).

File: modules/vaos/app/serializers/vaos/v2/eps_appointment_serializer.rb [hunk 13]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+module VAOS
+  module V2
+    class EpsAppointmentSerializer
```
Comment (kanchanasuriya) [inline modules/vaos/app/serializers/vaos/v2/eps_appointment_serializer.rb:5]: Could you please add tests for this serializer?

File: modules/vaos/spec/services/v2/appointment_service_spec.rb [hunk 14]
```diff
@@ -1052,6 +1111,20 @@
     end
   end
 
+  describe '#get_appointments merge' do
```
Comment (kanchanasuriya) [inline modules/vaos/spec/services/v2/appointment_service_spec.rb:1114]: Can we please add more tests for scenarios like no matching referral number, nil start date, empty eps_appointments or appointment data?

File: modules/vaos/app/services/vaos/v2/appointments_service.rb [hunk 15]
```diff
@@ -52,6 +52,11 @@ def get_appointments(start_date, end_date, statuses = nil, pagination_params = {
             cnp_count += 1 if cnp?(appt)
           end
 
+          if include[:eps]
+            # TODO: prepare eps_appointments differently than vaos appointments
```
Comment (kanchanasuriya) [inline modules/vaos/app/services/vaos/v2/appointments_service.rb:56]: Is this TODO still relevant for the changes here?

File: modules/vaos/app/serializers/vaos/v2/eps_appointment_serializer.rb (hunk 12 above)
Comment (lee-delarm6) [reply modules/vaos/app/serializers/vaos/v2/eps_appointment_serializer.rb]: This was a mistake, thanks for the catch. Causes issues with the validator (found it now that I wrote another test for duplicates)

**Review by stevenjcumming — CHANGES_REQUESTED** [review]
Please convert the EpsAppointmentSerializer to a non-Active Record model. 

Serializers should only be used to convert attributes (or a hash) to JSON for an api response.

File: modules/vaos/app/services/vaos/v2/appointments_service.rb (hunk 4 above)
Comment (lee-delarm6) [reply modules/vaos/app/services/vaos/v2/appointments_service.rb]: This has been updated, we're using referral_number, a few data points have been changed. We are also using `start` because referral_number is not purely unique, but it is when combined with a time

File: modules/vaos/app/serializers/vaos/v2/eps_appointment_serializer.rb [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+          id: appt[:id].to_s,
+          status: appt[:state] == 'booked',
+          patientIcn: appt[:patientId],
+          created: appt.dig(:appointmentDetails, :lastRetrieved),
+          requestedPeriods: prepare_requested_periods(appt[:appointmentDetails]),
+          locationId: appt[:locationId],
+          clinic: appt[:clinic],
+          start: appt[:start],
+          contact: appt[:contact],
+          referralID: appt.dig(:referral, :referralNumber),
```
[the remark this answers is not in the document]
Comment (lee-delarm6) [reply modules/vaos/app/serializers/vaos/v2/eps_appointment_serializer.rb]: I've dropped requested periods, especially since it is validated and blocks those appointments from showing up

File: modules/vaos/spec/services/v2/appointment_service_spec.rb [hunk 17]
```diff
@@ -44,6 +44,65 @@
...
+      {
+        id: 123,
+        state: 'submitted',
+        patientId: '456',
+        start: '2024-12-01T10:00:00Z',
+        appointmentDetails: {
+          lastRetrieved: '2024-12-01T10:00:00Z',
+          start: '2024-12-01T10:00:00Z'
+        },
+        locationId: '789',
```
[the remark this answers is not in the document]
Comment (lee-delarm6) [reply modules/vaos/spec/services/v2/appointment_service_spec.rb]: This schema has been updated quite a bit

File: modules/vaos/app/serializers/vaos/v2/eps_appointment_serializer.rb (hunk 13 above)
Comment (lee-delarm6) [reply modules/vaos/app/serializers/vaos/v2/eps_appointment_serializer.rb:5]: Will be in the next update

**Review by devin-mccurdy — DISMISSED** [review]
STILL VERY LGTM

File: config/settings.yml [hunk 18]
```diff
@@ -1703,6 +1703,7 @@ vaos:
     api_url: "https://api.wellhive.com"
     base_path: "care-navigation/v1"
     scopes: "care-nav"
+    fake_patient_id: "fake_patient_id"
```
[the remark this answers is not in the document]
Comment (lee-delarm6) [reply config/settings.yml:1706]: It's not needed at this point in time, I used it (in my local settings) to make real calls through the console to test some endpoints out. I'll remove it

**Review by randomsync — COMMENTED** [review]
looks good. mostly minor comments, except 1 that probably needs addressing

File: modules/vaos/app/serializers/vaos/v2/eps_appointment.rb [hunk 19]
```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+
+module VAOS
+  module V2
+    class EpsAppointment
```
[the remark this answers is not in the document]
Comment (lee-delarm6) [reply modules/vaos/app/serializers/vaos/v2/eps_appointment.rb:5]: I had that, but VFS team suggested this naming due to it not serializing any JSON data (directly)

File: modules/vaos/app/serializers/vaos/v2/eps_appointment.rb [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+          contact: @contact,
+          referral_id: @referral_id,
+          referral: @referral
+        }.compact
+      end
+
+      private
+
+      def determine_status(status)
+        status ? 'booked' : 'proposed'
```
[the remark this answers is not in the document]
Comment (lee-delarm6) [reply modules/vaos/app/serializers/vaos/v2/eps_appointment.rb:40]: Status is not generally true or false, but can not exist, which is where I was in my mindset initially writing this. It does make more sense to set it to a value if it has one, or "proposed" if it doesn't have "booked" specifically.

File: modules/vaos/app/serializers/vaos/v2/eps_appointment.rb [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+          contact: @contact,
+          referral_id: @referral_id,
+          referral: @referral
+        }.compact
+      end
+
+      private
+
+      def determine_status(status)
+        status || 'proposed'
```
Comment (randomsync) [inline modules/vaos/app/serializers/vaos/v2/eps_appointment.rb:40]: I don't think this is still quite right. `status` as a string will always be truthy, and will return status unless it's `nil`.

```
vets-api(dev)> 'asdf' || 'proposed'
=> "asdf"
vets-api(dev)> nil || 'proposed'
=> "proposed"
vets-api(dev)> 'not a valid status' || 'proposed'
=> "not a valid status"
```

If you want status to be 'proposed' unless it's specifically 'booked', this should work:
```suggestion
        status == 'booked' ? 'booked' : 'proposed'
```

File: modules/vaos/app/serializers/vaos/v2/eps_appointment.rb [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+
+module VAOS
+  module V2
+    class EpsAppointment
+      def initialize(params = {})
+        appointment_details = params[:appointment_details]
+        referral_details = params[:referral]
+
+        @id = params[:id]&.to_s
+        @status = appointment_details[:status],
```
Comment (randomsync) [inline modules/vaos/app/serializers/vaos/v2/eps_appointment.rb:11]: suggest removing the comma from here

**Review by randomsync — DISMISSED** [review]
lgtm with some minor feedback

**Comment by stevenjcumming** [comment]
@lee-delarm6 can you put this PR in a draft state until the lint/test warnings/failures are resolved?

**Comment by lee-delarm6** [comment]
> @lee-delarm6 can you put this PR in a draft state until the lint/test warnings/failures are resolved?

I'm updating to master branch, my tests should still be passing, I believe it's most likely due to not being updated to master

**Comment by lee-delarm6** [comment]
> @lee-delarm6 there is still a failing test

It was my test, I fixed it. I totally forgot I updated another file

**Review by stevenjcumming — CHANGES_REQUESTED** [review]
@lee-delarm6 the EpsAppointment class and corresponding spec should be moved under the `model/vaos/v2` folder

Sorry I didn't notice that sooner

**Review by randomsync — APPROVED** [review]

**Review by rmtolmach — APPROVED** [review]

**Review by rmtolmach — APPROVED** [review]
